File: src/Api/Services/UserService.cs
using System.Text.Json;
using Microsoft.EntityFrameworkCore;
using PostGenerator.Api.Data;
using PostGenerator.Api.Models;

namespace PostGenerator.Api.Services;

public class UserService : IUserService
{
    private readonly AppDbContext _db;
    private readonly IFirebaseAuthService _firebaseAuth;

    public UserService(AppDbContext db, IFirebaseAuthService firebaseAuth)
    {
        _db = db;
        _firebaseAuth = firebaseAuth;
    }

    public async Task<UserDto?> SyncFromTokenAsync(string idToken, CancellationToken cancellationToken = default)
    {
        var result = await _firebaseAuth.VerifyIdTokenAsync(idToken, cancellationToken);
        if (result == null) return null;

        var user = await _db.Users.FirstOrDefaultAsync(u => u.ExternalId == result.Uid, cancellationToken);
        if (user == null)
        {
            user = new User
            {
                ExternalId = result.Uid,
                Email = result.Email,
                Name = result.Name,
                CreatedAt = DateTime.UtcNow
            };
            _db.Users.Add(user);
        }
        else
        {
            user.Email = result.Email;
            user.Name = result.Name;
        }
        await _db.SaveChangesAsync(cancellationToken);
        return new UserDto(user.Id, user.Email, user.Name, user.PreferredTheme, user.AvatarUrl, user.CreatedAt);
    }

    public async Task<UserDto?> GetByIdAsync(int userId, CancellationToken cancellationToken = default)
    {
        var user = await _db.Users.AsNoTracking().FirstOrDefaultAsync(u => u.Id == userId, cancellationToken);
        return user == null ? null : new UserDto(user.Id, user.Email, user.Name, user.PreferredTheme, user.AvatarUrl, user.CreatedAt);
    }

    public async Task<bool> UpdateProfileAsync(int userId, UserProfileUpdateRequest request, CancellationToken cancellationToken = default)
    {
        var user = await _db.Users.FirstOrDefaultAsync(u => u.Id == userId, cancellationToken);
        if (user == null) return false;
        if (request.PreferredTheme is "light" or "dark")
            user.PreferredTheme = request.PreferredTheme;
        else if (request.PreferredTheme != null)
            user.PreferredTheme = null;
        if (request.AvatarUrl != null && request.AvatarUrl.Length > 2_000_000)
            throw new ArgumentException("Avatar URL must be 2,000,000 characters or less.", nameof(request));
        user.AvatarUrl = request.AvatarUrl;
        await _db.SaveChangesAsync(cancellationToken);
        return true;
    }

    public async Task<IReadOnlyList<SocialCredentialDto>> GetCredentialsAsync(int userId, CancellationToken cancellationToken = default)
    {
        var stored = await _db.UserSocialCredentials
            .AsNoTracking()
            .Where(c => c.UserId == userId)
            .ToDictionaryAsync(c => c.Platform, c => c.CredentialJson, StringComparer.OrdinalIgnoreCase, cancellationToken);
        var result = new List<SocialCredentialDto>();
        foreach (var platform in SocialPlatformKeys.AllPlatforms)
        {
            var keys = SocialPlatformKeys.KeysByPlatform.TryGetValue(platform, out var k) ? k : Array.Empty<string>();
            var masked = new Dictionary<string, string?>();
            if (stored.TryGetValue(platform, out var json))
            {
                var dict = JsonSerializer.Deserialize<Dictionary<string, string?>>(json) ?? new Dictionary<string, string?>();
                foreach (var key in keys)
                    masked[key] = dict.TryGetValue(key, out var v) && !string.IsNullOrEmpty(v) ? "***" : null;
            }
            else
            {
                foreach (var key in keys)
                    masked[key] = null;
            }
            result.Add(new SocialCredentialDto(platform, masked));
        }
        return result;
    }

    public async Task<bool> SetCredentialAsync(int userId, string platform, IReadOnlyDictionary<string, string?>? credentials, CancellationToken cancellationToken = default)
    {
        if (string.IsNullOrWhiteSpace(platform) || platform.Length > 32) return false;
        platform = platform.Trim();
        var existing = await _db.UserSocialCredentials.FirstOrDefaultAsync(c => c.UserId == userId && c.Platform == platform, cancellationToken);
        var dict = new Dictionary<string, string?>();
        if (existing != null)
        {
            dict = JsonSerializer.Deserialize<Dictionary<string, string?>>(existing.CredentialJson) ?? new Dictionary<string, string?>();
        }
        if (credentials != null)
        {
            foreach (var kv in credentials)
            {
                if (string.IsNullOrEmpty(kv.Value))
                    dict.Remove(kv.Key);
                else
                    dict[kv.Key] = kv.Value;
            }
        }
        var json = JsonSerializer.Serialize(dict);
        if (existing != null)
        {
            existing.CredentialJson = json;
            existing.UpdatedAt = DateTime.UtcNow;
        }
        else
        {
            _db.UserSocialCredentials.Add(new UserSocialCredential
            {
                UserId = userId,
                Platform = platform,
                CredentialJson = json,
                UpdatedAt = DateTime.UtcNow
            });
        }
        await _db.SaveChangesAsync(cancellationToken);
        return true;
    }
}

File: src/lib/services/user-service.ts
import { db } from "@/db/drizzle";
import { users, userSocialCredentials } from "@/db/schema";
import { eq, and } from "drizzle-orm";
import { ALL_PLATFORMS, KEYS_BY_PLATFORM } from "@/lib/social-platform-keys";

export interface UserProfile {
  id: number;
  email: string;
  name: string | null;
  preferredTheme: string | null;
  avatarUrl: string | null;
  createdAt: Date;
}

export interface SocialCredentialDto {
  platform: string;
  credentials: Record<string, string | null>;
}

export function toUserProfile(row: {
  id: number;
  email: string;
  name: string | null;
  preferredTheme: string | null;
  avatarUrl: string | null;
  createdAt: Date;
}): UserProfile {
  return {
    id: row.id,
    email: row.email,
    name: row.name,
    preferredTheme: row.preferredTheme,
    avatarUrl: row.avatarUrl,
    createdAt: row.createdAt,
  };
}

export async function getById(userId: number): Promise<UserProfile | null> {
  const [user] = await db
    .select()
    .from(users)
    .where(eq(users.id, userId))
    .limit(1);
  return user ? toUserProfile(user) : null;
}

export async function updateProfile(
  userId: number,
  updates: { preferredTheme?: string | null; avatarUrl?: string | null }
): Promise<boolean> {
  const [user] = await db
    .select()
    .from(users)
    .where(eq(users.id, userId))
    .limit(1);
  if (!user) return false;

  const next: Record<string, unknown> = {};
  if (updates.preferredTheme !== undefined) {
    next.preferredTheme =
      updates.preferredTheme === "light" || updates.preferredTheme === "dark"
        ? updates.preferredTheme
        : null;
  }
  if (updates.avatarUrl !== undefined) {
    if (updates.avatarUrl && updates.avatarUrl.length > 2_000_000) {
      throw new Error("Avatar URL must be 2,000,000 characters or less.");
    }
    next.avatarUrl = updates.avatarUrl;
  }

  if (Object.keys(next).length === 0) return true;

  await db.update(users).set(next).where(eq(users.id, userId));
  return true;
}

export async function getCredentials(
  userId: number
): Promise<SocialCredentialDto[]> {
  const stored = await db
    .select({ platform: userSocialCredentials.platform, credentialJson: userSocialCredentials.credentialJson })
    .from(userSocialCredentials)
    .where(eq(userSocialCredentials.userId, userId));

  const storedMap = new Map(
    stored.map((s) => [s.platform.toLowerCase(), s.credentialJson])
  );

  const result: SocialCredentialDto[] = [];
  for (const platform of ALL_PLATFORMS) {
    const keys = KEYS_BY_PLATFORM[platform] ?? [];
    const masked: Record<string, string | null> = {};
    const json = storedMap.get(platform.toLowerCase());
    if (json) {
      try {
        const dict = JSON.parse(json) as Record<string, string | null>;
        for (const key of keys) {
          masked[key] = dict[key] && dict[key]!.length > 0 ? "***" : null;
        }
      } catch {
        for (const key of keys) masked[key] = null;
      }
    } else {
      for (const key of keys) masked[key] = null;
    }
    result.push({ platform, credentials: masked });
  }
  return result;
}

export async function setCredential(
  userId: number,
  platform: string,
  credentials: Record<string, string | null> | null
): Promise<boolean> {
  const platformTrimmed = platform.trim();
  if (!platformTrimmed || platformTrimmed.length > 32) return false;

  const [existing] = await db
    .select()
    .from(userSocialCredentials)
    .where(
      and(
        eq(userSocialCredentials.userId, userId),
        eq(userSocialCredentials.platform, platformTrimmed)
      )
    )
    .limit(1);

  const dict: Record<string, string | null> = existing
    ? (JSON.parse(existing.credentialJson) as Record<string, string | null>)
    : {};

  if (credentials) {
    for (const [key, value] of Object.entries(credentials)) {
      if (!value) delete dict[key];
      else dict[key] = value;
    }
  }

  const json = JSON.stringify(dict);
  if (json.length > 8000) {
    throw new Error("Credential data exceeds maximum size of 8000 characters.");
  }

  const now = new Date();
  if (existing) {
    await db
      .update(userSocialCredentials)
      .set({ credentialJson: json, updatedAt: now })
      .where(eq(userSocialCredentials.id, existing.id));
  } else {
    await db.insert(userSocialCredentials).values({
      userId,
      platform: platformTrimmed,
      credentialJson: json,
      updatedAt: now,
    });
  }
  return true;
}
